### ISSUE TITLE
JUnit BOM should define fixed version ranges

### ISSUE THREAD

**Issue by opensource21** [junit-team/junit5#2353]
In a large projekt we made the experience that in some situation no test will run. It looks that this was illegal combination of platform version and junit version. I guess there exists only a few correct combinations. This end with junit 5.6 and runner 1.5 for example. So I think that in the junit-bom the should be a range [5.6.0, 5.7) and platform [1.6.0, 1.7).  Then maven should show explicitly the conflict.

**Comment by sbrannen** [comment]
Tentatively slated for 5.7 M2 for _team discussion_

**Comment by marcphilipp** [comment]
@opensource21 If I understand you correctly, you're saying Maven will emit a different error message if the BOM defined the versions as `[5.6.0, 5.7)` instead of `5.6.0`. Is that right?

**Comment by opensource21** [comment]
At the moment there will be no error-message because for maven it's fine to use different versions of the libraries. In the BOM only stands a wish or recommendation. There is no information that you shouldn't mix 5.6 with 5.7 and 1.7 runtime. But I didn't test it. It's just the way I would define it based on the maven-documentation if the mix can become critical (and I think this is the case). Unfortunately it's a relative complex szenario to reproduce the error.

**Comment by marcphilipp** [comment]
Our recommendation is to align the versions as specified in the BOM. Thus, I'm inclined to close this issue since I don't see how version ranges would improve the situation. Am I missing sth.?

**Comment by opensource21** [comment]
I think you miss something. If you start with https://github.com/junit-team/junit5-samples/blob/r5.6.2/junit5-jupiter-starter-maven/pom.xml.
You projekt grow, you make at dependcyManagement an import like
```
            <dependency>
                <groupId>org.springframework.boot</groupId>
                <artifactId>spring-boot-dependencies</artifactId>
                <version>2.2.4.RELEASE</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
```
You get mixed junit-versions and can have effects like no test will be executed. Perhaps you must further import archunit to get  really a problem.

**Comment by opensource21** [comment]
Here you see the result:
```
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ junit5-jupiter-starter-maven ---
[INFO] com.example:junit5-jupiter-starter-maven:jar:1.0-SNAPSHOT
[INFO] \- org.junit.jupiter:junit-jupiter:jar:5.6.2:test
[INFO]    +- org.junit.jupiter:junit-jupiter-api:jar:5.5.2:test
[INFO]    |  +- org.apiguardian:apiguardian-api:jar:1.1.0:test
[INFO]    |  +- org.opentest4j:opentest4j:jar:1.2.0:test
[INFO]    |  \- org.junit.platform:junit-platform-commons:jar:1.5.2:test
[INFO]    +- org.junit.jupiter:junit-jupiter-params:jar:5.5.2:test
[INFO]    \- org.junit.jupiter:junit-jupiter-engine:jar:5.5.2:test
[INFO]       \- org.junit.platform:junit-platform-engine:jar:1.5.2:test
```

**Comment by marcphilipp** [comment]
@opensource21 I acknowledge mixed versions are an issue, in particular when importing different Maven BOMs. However, I still fail to see how version ranges would help here since Maven will still just pick the first version it encounters, won't it?

**Comment by opensource21** [comment]
As far as I maven understand with the range it will ends in an error.  How ever I never create a test scenario. 
I'm unable to say how quick I will be able to create such a test scenario.
At least in the example you should import the bom in <dependencyManager>-Section. Then it's safe too and the solution we use in our projects.

**Comment by marcphilipp** [comment]
>  At least in the example you should import the bom in -Section. Then it's safe too and the solution we use in our projects.

What do you mean by that?

**Comment by opensource21** [comment]
In  https://github.com/junit-team/junit5-samples/blob/r5.6.2/junit5-jupiter-starter-maven/pom.xml. should be something like
```
    <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>org.junit</groupId>
                <artifactId>junit-bom</artifactId>
                <version>${junit-jupiter.version}</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
...
```

**Comment by marcphilipp** [comment]
Point taken, I will make the samples use the BOM.

@opensource21 Could you please create the test scenario you had in mind?

**Comment by khmarbaise** [comment]
You must be aware that if you are using a BOM import for example from spring boot you will end up with the version for JUnit Jupiter which is defined by spring boot. If you like to have a different version you simply have to do the bom import before spring boot.

**Comment by opensource21** [comment]
@khmarbaise Can you provide more details, about what's wrong in the Orcale-doku? What is a better document? Why not use version ranges?

**Comment by khmarbaise** [comment]
Simply more or less in everything cause it relates to Maven 2: ([Maven 3 is out for 10 years](https://maven.apache.org/docs/history.html))

See some details https://maven.apache.org/pom.html#version-order-specification

> Identical versions with different qualifier fields are compared by using basic string comparison.
> 
> For example:
> 
> 1.2-beta-2 is newer than 1.2-alpha-6.

```
java -jar ~/tools/apache-maven-3.6.3/lib/maven-artifact-3.6.3.jar 1.2-beta-2 1.2-alpha-6
Display parameters as parsed by Maven (in canonical form) and comparison result:
1. 1.2-beta-2 == 1.2-beta-2
   1.2-beta-2 > 1.2-alpha-6
2. 1.2-alpha-6 == 1.2-alpha-6
```
Which is correct. 
Furthermore the claimed ordering in the article:
```
java -jar ~/tools/apache-maven-3.6.3/lib/maven-artifact-3.6.3.jar 1.0.9.3 1.0.10.1
Display parameters as parsed by Maven (in canonical form) and comparison result:
1. 1.0.9.3 == 1.0.9.3
   1.0.9.3 < 1.0.10.1
2. 1.0.10.1 == 1.0.10.1
```
which showed that maven handles that correctly. `1.0.9.3` is before `1.0.10.1`. ([see my detailed blog post about it](https://blog.soebes.de/blog/2017/02/04/apache-maven-how-version-comparison-works/).
You can use as much numbers as you like `1.3.4.56.7.8.9.5.3.2.4` and it will be compared correctly. 
Furthermore it's mixing external plugins (maven versions plugin which is not maven versions plugin it is [`versions-maven-plugin`](https://github.com/mojohaus/versions-maven-plugin)  instead) and Maven itself.
A statement: `When Maven encounters multiple matches for a version reference, it uses the highest matching version.` is simply wrong cause Maven handles based on distance not on version. If you are using version ranges than the selected version is the highest but not for a dependency.

Version ranges make your build not reproducible. Today you can build while you get a version lets say 1.0.1 for a dependency and you build the same project tomorrow and a newer version is available (like 1.0.2). You might get different results.

**Comment by marcphilipp** [comment]
@khmarbaise Thanks for providing those insights!

In the light of the above discussion, I'm closing this issue.

**Comment by opensource21** [comment]
@khmarbaise I don't see what's wrong with the information about ranges in the oracle-article? Furthermore I don't find additional information about how to define ranges in your linked article.
I'm agree that ranges are suboptimal, but the current -situation ends in projects where we can have mixed junit-versions which doesn't work together, so that at the end of the day no test will executed - a nightmare. In great projects this isn't easy to detect.
Found a lot of hint's at stackoverflow which recommend to add a lot of junit5-jars directly which I think is wrong. The cleanest way is to import the bom at first at the dependencyManagement-Section. 

I thought that maven was able to define dependencies in a way the prevent from versions mixes which doesn't work. But we can leave this issue close - I have learned how to handle the situation. Still think that a technical solution is possible.
